From 6eef2e5a3af17ca95ef1972c94349e4c3b268274 Mon Sep 17 00:00:00 2001
From: wuzhouhui <wuzhouhui250@gmail.com>
Date: Sat, 2 Mar 2019 21:47:19 +0800
Subject: [PATCH 92/95] svn/ci: add option --bypass-hooks to skip all client
 side hooks

The name of option come from --bypass-hooks of svnadmin
---
 subversion/svn/cl.h               | 1 +
 subversion/svn/commit-cmd.c       | 5 ++++-
 subversion/svn/svn.c              | 8 +++++++-
 tools/client-side/bash_completion | 2 +-
 4 files changed, 13 insertions(+), 3 deletions(-)

diff --git a/subversion/svn/cl.h b/subversion/svn/cl.h
index 1e167b3..256dbbb 100644
--- a/subversion/svn/cl.h
+++ b/subversion/svn/cl.h
@@ -263,6 +263,7 @@ typedef struct svn_cl__opt_state_t
   svn_boolean_t adds_as_modification; /* update 'add vs add' no tree conflict */
   svn_boolean_t vacuum_pristines; /* remove unreferenced pristines */
   svn_boolean_t list;
+  svn_boolean_t bypass_hooks;      /* bypass client side hooks */
 } svn_cl__opt_state_t;
 
 /* Conflict stats for operations such as update and merge. */
diff --git a/subversion/svn/commit-cmd.c b/subversion/svn/commit-cmd.c
index 9f09d62..678d12d 100644
--- a/subversion/svn/commit-cmd.c
+++ b/subversion/svn/commit-cmd.c
@@ -172,7 +172,10 @@ svn_cl__commit(apr_getopt_t *os,
   svn_opt_push_implicit_dot_target(targets, pool);
 
   /* Execute pre-commit hook if possible */
-  SVN_ERR(pre_commit(ctx, APR_ARRAY_IDX(targets, 0, const char *), pool));
+  if (opt_state->bypass_hooks == FALSE)
+    {
+      SVN_ERR(pre_commit(ctx, APR_ARRAY_IDX(targets, 0, const char *), pool));
+    }
 
   SVN_ERR(svn_cl__eat_peg_revisions(&targets, targets, pool));
 
diff --git a/subversion/svn/svn.c b/subversion/svn/svn.c
index 7fb07ee..9bbcc56 100644
--- a/subversion/svn/svn.c
+++ b/subversion/svn/svn.c
@@ -73,6 +73,7 @@ typedef enum svn_cl__longopt_t {
   opt_auth_password_from_stdin,
   opt_auth_username,
   opt_autoprops,
+  opt_bypass_hooks,
   opt_changelist,
   opt_config_dir,
   opt_config_options,
@@ -494,6 +495,7 @@ const apr_getopt_option_t svn_cl__options[] =
   {"list", opt_list, 0, N_("list shelved patches")},
   {"keep-shelved", opt_keep_shelved, 0, N_("do not delete the shelved patch")},
   {"delete", opt_delete, 0, N_("delete the shelved patch")},
+  {"bypass-hooks", opt_bypass_hooks, 0, N_("Bypass client side hooks\n")},
 
   /* Long-opt Aliases
    *
@@ -692,7 +694,9 @@ const svn_opt_subcommand_desc2_t svn_cl__cmd_table[] =
        "  externals reached by recursion. Do not commit externals with a\n"
        "  fixed revision.\n"),
     {'q', 'N', opt_depth, opt_targets, opt_no_unlock, SVN_CL__LOG_MSG_OPTIONS,
-     opt_changelist, opt_keep_changelists, 'v', opt_include_externals},
+     opt_changelist, opt_keep_changelists, 'v', opt_include_externals,
+     opt_bypass_hooks
+    },
     { 'v', N_("show unified diff that this commit will produce") } },
 
   { "copy", svn_cl__copy, {"cp"}, N_
@@ -2735,6 +2739,8 @@ sub_main(int *exit_code, int argc, const char *argv[], apr_pool_t *pool)
       case opt_vacuum_pristines:
         opt_state.vacuum_pristines = TRUE;
         break;
+      case opt_bypass_hooks:
+        opt_state.bypass_hooks = TRUE;
       default:
         /* Hmmm. Perhaps this would be a good place to squirrel away
            opts that commands like svn diff might need. Hmmm indeed. */
diff --git a/tools/client-side/bash_completion b/tools/client-side/bash_completion
index bc23ca3..0fd353c 100644
--- a/tools/client-side/bash_completion
+++ b/tools/client-side/bash_completion
@@ -891,7 +891,7 @@ _svn()
 	commit|ci)
 		cmdOpts="$mOpts $qOpts $nOpts --targets --editor-cmd $pOpts \
 		         --no-unlock $cOpts --keep-changelists \
-		         --include-externals -v --verbose"
+		         --include-externals -v --verbose --bypass-hooks"
 		;;
 	copy|cp)
 		cmdOpts="$mOpts $rOpts $qOpts --editor-cmd $pOpts --parents \
-- 
1.8.3.1

